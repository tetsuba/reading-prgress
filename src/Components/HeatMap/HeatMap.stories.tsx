import type { Meta, StoryObj } from '@storybook/react'
import HeatMap from './HeatMap'

/**
 * the HeatMap component generates a visual colored keyword grid to represent
 * relative data values for a set of words.
 *
 */
const meta = {
    title: 'COMPONENTS/HeatMap',
    component: HeatMap,
    parameters: {
        layout: 'centered'
    },
    tags: ['autodocs'],
    args: {},
    argTypes: {}
} satisfies Meta<typeof HeatMap>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
    args: {
        children: 'Header',
        words: [
            { word: 'word', index: 1 },
            { word: 'that', index: 2 },
            { word: 'that', index: 3 },
            { word: 'that', index: 1 },
            { word: 'that', index: 4 },
            { word: 'that', index: 7 },
            { word: 'that', index: 9 }
        ],
        search: '',
        color: 'red'
    }
}
