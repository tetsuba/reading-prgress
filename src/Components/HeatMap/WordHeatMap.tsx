import { HeatMapColors } from './HeatMap'

export type WordHeatMapTypes = {
    word: string
    index: number
    max: number
    color: HeatMapColors
}

export default function WordHeatMap(props: WordHeatMapTypes) {
    const rgb = 255 * (props.index / props.max)
    const numberHeatMap = {
        backgroundColor: `rgb(${255 - rgb},${255 - rgb},${255 - rgb})`
    }
    const showBadge = props.index > 1

    function getBackGroundStyles() {
        if (props.color === 'red') {
            return { backgroundColor: `rgb(255,${255 - rgb},${255 - rgb})` }
        }
        if (props.color === 'blue') {
            return {
                backgroundColor: `rgb(${255 - rgb}, ${255 - rgb / 2}, 255)`
            }
        }
        if (props.color === 'green') {
            return {
                backgroundColor: `rgb(${255 - rgb}, ${255 - rgb / 3}, ${
                    255 - rgb
                })`
            }
        }
        if (props.color === 'none') {
            return { backgroundColor: 'rgb(230,230,230)' }
        }
    }

    function getTextStyles() {
        if (props.color === 'none') {
            return 'text-gray-900'
        }
        return (props.index / props.max) * 100 > 50
            ? 'text-white'
            : 'text-gray-900'
    }

    return (
        <span
            style={getBackGroundStyles()}
            data-testid="heat-map-word"
            className={`m-2 rounded-2xl border-2 p-3 md:m-5`}
        >
            <span className={`font-bold ${getTextStyles()}`}>{props.word}</span>
            {showBadge && (
                <span
                    style={numberHeatMap}
                    className={`ml-1 rounded-full py-1 px-2 text-white md:ml-3`}
                >
                    {props.index}
                </span>
            )}
        </span>
    )
}
