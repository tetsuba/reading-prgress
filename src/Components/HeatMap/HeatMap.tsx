import WordHeatMap from './WordHeatMap'
import H3 from '../H3/H3'

export type HeatMapColors = 'red' | 'blue' | 'green' | 'none'

type WordTypes = {
    word: string
    index: number
}

type PropTypes = {
    words: WordTypes[]
    search: string
    children: string
    color: HeatMapColors
}
export default function HeatMap(props: PropTypes) {
    const max = props.words.map(({ index }) => index).sort((a, b) => b - a)[0]

    function sightWords(words: WordTypes) {
        const regEx = new RegExp(props.search.toLowerCase(), 'g')
        return regEx.test(words.word.toLowerCase())
    }

    function heatMap(words: WordTypes, i: number) {
        return (
            <WordHeatMap
                color={props.color}
                max={max}
                key={`word-${i}`}
                word={words.word}
                index={words.index}
            />
        )
    }

    return (
        <>
            <H3 className="mt-10 rounded-t-lg bg-gray-200 p-2">
                {props.children}
            </H3>
            <div
                data-testid="heat-map"
                className="min-h-96 md:border-3 inline-flex w-full flex-wrap items-center justify-center rounded-b-lg border-2 border-t-0 border-dashed border-gray-200 p-2 shadow-md"
            >
                {props.words.filter(sightWords).map(heatMap)}
            </div>
        </>
    )
}
