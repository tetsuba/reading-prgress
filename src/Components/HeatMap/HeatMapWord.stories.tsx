import type { Meta, StoryObj } from '@storybook/react'
import WordHeatMap from './WordHeatMap'

const meta = {
    title: 'COMPONENTS/WordHeatMap',
    component: WordHeatMap,
    parameters: {
        layout: 'centered'
    },
    tags: ['autodocs'],
    args: {},
    argTypes: {}
} satisfies Meta<typeof WordHeatMap>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
    args: {
        max: 10,
        color: 'red',
        data: { word: 'this', index: 5 },
        index: 2
    }
}
