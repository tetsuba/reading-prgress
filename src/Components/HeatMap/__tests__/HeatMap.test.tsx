import { render } from '@testing-library/react'
import HeatMap from '../HeatMap'

describe('HeatMap', () => {
    test('should render a word heatmap', () => {
        const words = [
            { word: 'there', index: 1 },
            { word: 'then', index: 5 }
        ]
        const { asFragment } = render(
            <HeatMap words={words} search="" color="red">
                Test Header
            </HeatMap>
        )
        expect(asFragment()).toMatchSnapshot()
    })

    test('should render nothing if words undefined', () => {
        const words = undefined
        const { asFragment } = render(
            <HeatMap words={words} search="" color="red">
                Test Header
            </HeatMap>
        )
        expect(asFragment()).toMatchSnapshot()
    })
})
