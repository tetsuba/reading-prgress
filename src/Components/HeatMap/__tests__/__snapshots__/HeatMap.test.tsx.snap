// Vitest Snapshot v1

exports[`HeatMap > should render a word heatmap 1`] = `
<DocumentFragment>
  <h3
    class="text-ml font-medium text-gray-900 dark:text-white md:text-xl mt-10 rounded-t-lg bg-gray-200 p-2"
  >
    Test Header
  </h3>
  <div
    class="min-h-96 md:border-3 inline-flex w-full flex-wrap items-center justify-center rounded-b-lg border-2 border-t-0 border-dashed border-gray-200 p-2 shadow-md"
    data-testid="heat-map"
  >
    <span
      class="m-2 rounded-2xl border-2 p-3 md:m-5"
      data-testid="heat-map-word"
      style="background-color: rgb(255, 204, 204);"
    >
      <span
        class="font-bold text-gray-900"
      >
        there
      </span>
    </span>
    <span
      class="m-2 rounded-2xl border-2 p-3 md:m-5"
      data-testid="heat-map-word"
      style="background-color: rgb(255, 0, 0);"
    >
      <span
        class="font-bold text-white"
      >
        then
      </span>
      <span
        class="ml-1 rounded-full py-1 px-2 text-white md:ml-3"
        style="background-color: rgb(0, 0, 0);"
      >
        5
      </span>
    </span>
  </div>
</DocumentFragment>
`;
